src/moonlink_service/src/error.rs

```diff
@@ -1,15 +1,118 @@
-#[derive(Debug, thiserror::Error)]
+use arrow_schema::ArrowError;
+use moonlink_error::{ErrorStatus, ErrorStruct};
+use std::io;
+use std::panic::Location;
+use std::result;
+use std::sync::Arc;
+use thiserror::Error;
+
+#[derive(Clone, Debug, Error)]
 pub enum Error {
-    #[error("Arrow schema error: {0}")]
-    ArrowSchema(#[from] arrow_schema::ArrowError),
-    #[error("Backend error: {0}")]
-    Backend(#[from] moonlink_backend::Error),
-    #[error("IO error: {0}")]
-    Io(#[from] std::io::Error),
-    #[error("RPC error: {0}")]
-    Rpc(#[from] moonlink_rpc::Error),
-    #[error("Join error: {0}.")]
-    TokioTaskJoin(#[from] tokio::task::JoinError),
-}
-
-pub type Result<T> = std::result::Result<T, Error>;
+    #[error("{0}")]
+    Arrow(ErrorStruct),
+
+    #[error("{0}")]
+    Backend(ErrorStruct),
+
+    #[error("{0}")]
+    Io(ErrorStruct),
+
+    #[error("{0}")]
+    Rpc(ErrorStruct),
+
+    #[error("{0}")]
+    TaskJoin(ErrorStruct),
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+impl From<ArrowError> for Error {
+    #[track_caller]
+    fn from(source: ArrowError) -> Self {
+        let status = match source {
+            ArrowError::IoError(_, _) => ErrorStatus::Temporary,
+
+            // All other errors are regard as permanent
+            _ => ErrorStatus::Permanent,
+        };
+
+        Error::Arrow(ErrorStruct {
+            message: format!("Arrow error: {source}"),
+            status,
+            source: Some(Arc::new(source.into())),
+            location: Some(Location::caller()),
+        })
+    }
+}
+
+impl From<moonlink_backend::Error> for Error {
+    #[track_caller]
+    fn from(source: moonlink_backend::Error) -> Self {
+        Error::Backend(ErrorStruct {
+            message: format!("Backend error: {source}"),
+            status: ErrorStatus::Permanent,
```

Comment:
Backend errors should have `ErrorStruct` inside, so whether permanent/transient should inherit from internal error